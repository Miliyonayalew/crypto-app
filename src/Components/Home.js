import React, { useEffect, useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Spinner } from 'loading-animations-react';
import Navbar from './Navbar';
import {
  fetchCoins, error, loading,
  allcrypto,
} from '../redux/features/coinSlice';
import Card from './Card';

let didInit = false;
const Home = () => {
  const dispatch = useDispatch();
  const coins = useSelector(allcrypto);
  const Loading = useSelector(loading);
  const Error = useSelector(error);

  const [search, setSearch] = useState('');
  const handleSearch = (e) => {
    setSearch(e.target.value);
  };

  useEffect(() => {
    if (!didInit) {
      dispatch(fetchCoins());
      didInit = true;
    }
  }, [dispatch]);

  const filteredCoins = coins ? coins.filter(
    (coin) => coin.name
      .toLowerCase()
      .includes(
        search.toLowerCase()
              || coin.symbol.toLowerCase().includes(search.toLowerCase()),
      ) || coin.rank.toString().includes(search.toString()),
  ) : [];
  return (
    <div>
      <Navbar title="Crypto App" />
      <div className="stat-container">
        <h1 className="title">Crypto Stats</h1>
        <div className="stats d-flex">
          <div>
            <h2>Market Cap</h2>
            <p>
              {coins[0]?.marketCap}
              {' '}
              B
            </p>
          </div>
          <div>
            <h2>24h Volume</h2>
            <p>
              {coins[0]?.volume}
              {' '}
              B
            </p>
          </div>
        </div>
      </div>
      <div className="container d-flex">
        <div className="search-container d-flex">
          <input
            type="text"
            placeholder="Search by name, symbol or rank"
            className="search"
            onChange={handleSearch}
            value={search}
          />
        </div>
        <div className="d-flex spinner">
          {Loading && <Spinner />}
          {Error && <h1>{Error}</h1>}
        </div>
        <div className="card-container">
          {filteredCoins.map((coin) => (
            <Card
              key={coin.id}
              coin={coin}
            />
          ))}
        </div>
      </div>
    </div>
  );
};

export default Home;
