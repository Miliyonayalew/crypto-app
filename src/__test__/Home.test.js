import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import store from '../redux/configureStore';
import Home from '../Components/Home';

describe('Test the Components', () => {
  test('Renders Home correctly', () => {
    const home = render(
      <Provider store={store}>
        <Router>
          <Home />
        </Router>
      </Provider>,
    );
    expect(home).toMatchSnapshot();
  });
  test('Renders Home with correct text', () => {
    const home = render(
      <Provider store={store}>
        <Router>
          <Home />
        </Router>
      </Provider>,
    );
    expect(home.getByText('Crypto Stats')).toBeInTheDocument();
    expect(home.getByText('Market Cap')).toBeInTheDocument();
    expect(home.getByText('24h Volume')).toBeInTheDocument();
    expect(home.getByText('Top 100 Cryptocurrencies')).toBeInTheDocument();
  });
});
