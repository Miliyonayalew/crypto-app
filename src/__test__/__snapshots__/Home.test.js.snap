// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the Components Renders Home correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app-container"
      >
        <div
          class="header d-flex"
          data-testid="testNavbar"
        >
          <a
            aria-current="page"
            class="navLink d-flex active"
            href="/"
          >
            <img
              alt="logo"
              src="logo.png"
            />
            Crypto App
          </a>
          <div
            class="icons d-flex"
          >
            <svg
              class="icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 14c1.66 0 2.99-1.34 2.99-3L15 5c0-1.66-1.34-3-3-3S9 3.34 9 5v6c0 1.66 1.34 3 3 3zm5.3-3c0 3-2.54 5.1-5.3 5.1S6.7 14 6.7 11H5c0 3.41 2.72 6.23 6 6.72V21h2v-3.28c3.28-.48 6-3.3 6-6.72h-1.7z"
              />
            </svg>
            <svg
              class="icon"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="3"
              />
              <path
                d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
              />
            </svg>
          </div>
        </div>
        <div
          class="stat-container"
        >
          <h1
            class="title"
          >
            Crypto Stats
          </h1>
          <div
            class="stats d-flex"
          >
            <img
              alt="logo"
              src="logo.png"
            />
            <div
              class="stats-data d-flex"
            >
              <div>
                <h2>
                  Market Cap
                </h2>
                <p>
                   
                  B
                </p>
              </div>
              <div>
                <h2>
                  24h Volume
                </h2>
                <p>
                   
                  B
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="container d-flex"
        >
          <div
            class="search-container d-flex"
          >
            <input
              class="search"
              placeholder="Search by name, symbol or rank"
              type="text"
              value=""
            />
          </div>
          <h2>
            Top 100 Cryptocurrencies
          </h2>
          <div
            class="d-flex spinner"
          >
            <div
              class="_index__spinnerWrapper "
              id=""
            >
              <div
                class="_index__spinner"
              >
                <svg
                  class="_index__viewBox"
                  viewBox="0 0 50 50"
                >
                  <circle
                    class="_index__circle"
                    cx="25"
                    cy="25"
                    fill="none"
                    r="22.5"
                    stroke="url(#myGradient)"
                    stroke-width="5"
                  />
                  <defs>
                    <lineargradient
                      id="myGradient"
                    >
                      <stop
                        offset="0%"
                        stop-color="green"
                      />
                      <stop
                        offset="100%"
                        stop-color="lightGreen"
                      />
                    </lineargradient>
                  </defs>
                </svg>
                <div
                  class="gradientText _index__spinnerText"
                >
                  Loading...
                </div>
              </div>
              <style>
                ._index__spinnerText {
    background: linear-gradient(90deg, green, lightGreen);
    background-clip: text;
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    font-weight: bold;
  }
              </style>
            </div>
          </div>
          <div
            class="card-container"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="app-container"
    >
      <div
        class="header d-flex"
        data-testid="testNavbar"
      >
        <a
          aria-current="page"
          class="navLink d-flex active"
          href="/"
        >
          <img
            alt="logo"
            src="logo.png"
          />
          Crypto App
        </a>
        <div
          class="icons d-flex"
        >
          <svg
            class="icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M12 14c1.66 0 2.99-1.34 2.99-3L15 5c0-1.66-1.34-3-3-3S9 3.34 9 5v6c0 1.66 1.34 3 3 3zm5.3-3c0 3-2.54 5.1-5.3 5.1S6.7 14 6.7 11H5c0 3.41 2.72 6.23 6 6.72V21h2v-3.28c3.28-.48 6-3.3 6-6.72h-1.7z"
            />
          </svg>
          <svg
            class="icon"
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="12"
              r="3"
            />
            <path
              d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
            />
          </svg>
        </div>
      </div>
      <div
        class="stat-container"
      >
        <h1
          class="title"
        >
          Crypto Stats
        </h1>
        <div
          class="stats d-flex"
        >
          <img
            alt="logo"
            src="logo.png"
          />
          <div
            class="stats-data d-flex"
          >
            <div>
              <h2>
                Market Cap
              </h2>
              <p>
                 
                B
              </p>
            </div>
            <div>
              <h2>
                24h Volume
              </h2>
              <p>
                 
                B
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="container d-flex"
      >
        <div
          class="search-container d-flex"
        >
          <input
            class="search"
            placeholder="Search by name, symbol or rank"
            type="text"
            value=""
          />
        </div>
        <h2>
          Top 100 Cryptocurrencies
        </h2>
        <div
          class="d-flex spinner"
        >
          <div
            class="_index__spinnerWrapper "
            id=""
          >
            <div
              class="_index__spinner"
            >
              <svg
                class="_index__viewBox"
                viewBox="0 0 50 50"
              >
                <circle
                  class="_index__circle"
                  cx="25"
                  cy="25"
                  fill="none"
                  r="22.5"
                  stroke="url(#myGradient)"
                  stroke-width="5"
                />
                <defs>
                  <lineargradient
                    id="myGradient"
                  >
                    <stop
                      offset="0%"
                      stop-color="green"
                    />
                    <stop
                      offset="100%"
                      stop-color="lightGreen"
                    />
                  </lineargradient>
                </defs>
              </svg>
              <div
                class="gradientText _index__spinnerText"
              >
                Loading...
              </div>
            </div>
            <style>
              ._index__spinnerText {
    background: linear-gradient(90deg, green, lightGreen);
    background-clip: text;
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    font-weight: bold;
  }
            </style>
          </div>
        </div>
        <div
          class="card-container"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
